### PR TITLE
feat (invoices): add possibility to fetch and filter invoices

### PR DESCRIPTION
**Pull request by lovrocolic**
## Context

Provide possibility to fetch and filter invoices

## Description

Add filter by payment status in the GET /invoices endpoint
Add resolver to fetch invoices and filter them by payment status

### CODE REVIEW

File: spec/graphql/resolvers/invoices_resolver_spec.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+      expect(invoices_response['metadata']['currentPage']).to eq(1)
+      expect(invoices_response['metadata']['totalCount']).to eq(2)
+    end
+  end
+
+  context 'when filtering by succeeded payment status' do
+    let(:query) do
+      <<~GQL
+      query {
```
Comment (jdenquin) [inline spec/graphql/resolvers/invoices_resolver_spec.rb:52]: Layout/HeredocIndentation: Use 2 spaces for indentation in a heredoc. (https://rubystyle.guide#squiggly-heredocs)

File: spec/requests/api/v1/invoices_spec.rb [hunk 2]
```diff
@@ -134,5 +134,24 @@
...
+      let(:invoice2) { create(:invoice, customer: customer, payment_status: :failed) }
+      let(:invoice3) { create(:invoice, customer: customer, payment_status: :pending) }
+
+      before do
+        invoice2
+        invoice3
+      end
+
+      it 'returns invoices with correct payment status' do
+        get_with_token(organization, "/api/v1/invoices?payment_status=pending")
```
Comment (jdenquin) [inline spec/requests/api/v1/invoices_spec.rb:149]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

**Review by rsempe — APPROVED** [review]

File: app/controllers/api/v1/invoices_controller.rb [hunk 3]
```diff
@@ -31,6 +31,11 @@ def index
         if params[:external_customer_id]
           invoices = invoices.joins(:customer).where(customers: { external_id: params[:external_customer_id] })
         end
+
+        if valid_payment_status?(params[:payment_status])
```
Comment (vincent-pochet) [inline app/controllers/api/v1/invoices_controller.rb:35]: Wondering if all this logic should be part of a query class:

```ruby
class BaseQuery < BaseService
  def initialize(organization)
    @organization = organization

    super
  end

  private

  attr_reader :organization
end

class InvoiceQuery < BaseQuery
  def call(filters)
    invoices = base_scope.xxx

    result.invoices = invoices
    result
  end

  private

  def base_scope
    organization.invoices
  end
end
```

That way you can even handle error with filters, or have shared methods for pagination

In the controller:

```ruby
def index
  query = InvoiceQuery.new(current_organization)
  result = query.call(params.permit(:external_customer_id, :payment_status))

  if result.success?
    render(
        json: ::CollectionSerializer.new(
          invoices,
          ::V1::InvoiceSerializer,
          collection_name: 'invoices',
          meta: pagination_metadata(invoices),
        ),
      )
  else
    render_error_response(result)
  end
end
```

**Review by vincent-pochet — APPROVED** [review]
Looks good, but since we will have more and more filter logic in controller and in GraphQL resolvers, I wonder if we shouldn't find a way to share and delegate some logic.

Note: it could probably wait for a specific PR ;)

File: app/controllers/api/v1/invoices_controller.rb (hunk 3 above)
Comment (lovrocolic) [reply app/controllers/api/v1/invoices_controller.rb:35]: opened a card for it :)

**Pull request merged**